### ISSUE TITLE
Improve methods to query hashed keys

### ISSUE THREAD

**Issue by nickReyn** [TykTechnologies/tyk#1256]
## The problem 
When creating keys with hashing turned on, our users can't query those keys without knowing the original key. Additionally, hash can't be obtained until the key is used, so queries such as the following won't work:
* Show me all keys globally
* Show me all keys we have per API/policy

## Potential solution
- Allow listing hashed keys
- When creating a Key, the hash is returned and the user can choose to record this using their own method. The hash can then be used to query keys usage, without exposing the key itself.

## Community feedback
We are keen to understand who else of the Tyk community has issues querying hashed keys and their suggested solutions! 👍

**Comment by lonelycode** [comment]
> When creating a Key, the hash is returned and the user can choose to record this using their own method. The hash can then be used to query keys usage, without exposing the key itself.

You can enable an event hook on the gateway that can send a webhook or run a plugin whenever a key is created. This will include the *raw original token in the payload*, the user can then use this to store the token in a secure location if they need to. This is already possible, we may want to improve the documentation around it.

**Comment by nextmetaphor** [comment]
When key hashing is enabled, for the [gateway API](https://tyk.io/docs/tyk-rest-api/token-management/), it would be great if...

- `/tyk/keys/create` `[POST]` also returned the key hash

- `/tyk/keys/{access_token}` `[GET]` could also accept the key hash if the `hashed` boolean parameter is set, as is currently supported for the `DELETE` method. This obviously wouldn't return the key value itself, but would be used to examine what access a key has access to, rate limits, policies etc from the hash of the key

- `/tyk/keys/` `[GET]` would return the key hash for all keys

- `/tyk/keys/` `[GET]` with a string `policy_id` parameter would return all hashed keys associated with the specified policy

- `/tyk/keys/` `[GET]` with a string `api_id` parameter would return all hashed keys associated with the specified api

This functionality would also be replicated on the dashboard API.

In essence the above allows an administrator of the gateway, who doesn't know (or need to know) the actual key itself, to ensure that the current keys are auditable, and can be removed if required.

**Comment by lonelycode** [comment]
We've had a discussion internally and are going to adopt parts of the recommendation proposed by @nextmetaphor - in particular the improvements to the existing Tyk API.

With regards to the filtering options, these look like they will be superseded by a different feature which is currently in planning.

**Comment by nickReyn** [comment]
@buger @lghiur can we discuss the UX/UI impact of this please?

**Comment by nickReyn** [comment]
This release: backend and Dashboard API, next release (2.7) can do front-end

**Comment by buger** [comment]
@lonelycode plan looks sane. The only addition that listing hashed keys should be explicitly allowed in tyk config, and disabled by default, since it can be safely enabled only for single org installations.

**Comment by buger** [comment]
For the Dashboard:
- add an additional checkbox "Hashed key" on API search screen.
- when viewing hashed key info, the form should be read-only (updates be protected on API level as well)

**Comment by buger** [comment]
@dencoded assigning to you, pls implement GW and Dashboard API side.

**Comment by lghiur** [comment]
@dencoded @nickReyn @buger is there any FE task opened for this? How the FE change should look exactly?

**Comment by nickReyn** [comment]
No FE task for this that I know of. And I don't think Ui has been defined. Sorry about that, never got round to looking at it. Suggest we move FE to 2.7 as I have commercials priority this week. Unless you want to take a look @ViolaMarku ?

**Comment by buger** [comment]
Well, in fact, changes needed to be done are very simple, the first stage can be just adding an additional checkbox on API search screen, called "Hashed key". When turned on, and when you put key id/hash to search box, it will add "hashed=true" query param when calling /apis/:apiId/keys/:keyId endpoint.

The rest of changes, when we need to disallow hashed key updates and etc, can be done via server side errors.

**Comment by buger** [comment]
@ConsM can you make this change?

**Comment by ConsM** [comment]
@buger sure

**Comment by nickReyn** [comment]
![tyk io api gateway 5](https://user-images.githubusercontent.com/15945383/37825073-984e6b6e-2e86-11e8-9d8d-753f819e91c1.png)

**Comment by bitsofinfo** [comment]
I was just made aware of this ticket, the key things I'd like to see are:

a) When in log browser, under gateway metadata -> tags. The "key-XXXX" listed there should be linked to a view of the attached session. Any "tags" in the session attached to that key should be listed there as well. (under gateway metadata -> tags) and ideally the alias of that key

b) Under "key management" I should be able to paste in the XXXX hashed portion of the "key-XXXX" referenced in log-browser, actually find it and see information attached to that key (session info, aliases, tags etc)

c) In activity by key, the search listing lists the hash + alias (great!), when I click into the graph view for activity, there is not alias presented. This details screen should show all session details attached to that key as well as the activity data.

**Comment by buger** [comment]
This change covers b). 

a) Log browser already should list all tags attached to the current session. The logic here is that analytics record tags field, which is displayed in UI is constructed by appending all the following tags below:
1) key tags
2) custom header tags - https://tyk.io/docs/tyk-rest-api/api-definition-objects/custom-analytics/
3) Config related tags  (for example all hybrid requests have `tyk-hybrid-rpc` tag)
4) key-X, org-X and api-X

@nickReyn do you remember if we have tasks for making key hashed clickable in log browser? and for c) ?

**Comment by bitsofinfo** [comment]
So for (A) - in log-browser I am not seeing any of the session tags attached to a key listed under "gateway metadata --> tags"

**Comment by nickReyn** [comment]
@buger no we do not have a ticket for it that i'm aware of, we can create one.

c) we have this in under 2.7 milestone which should cover the above - https://github.com/TykTechnologies/tyk/issues/1556 (@bitsofinfo pls add more to the ticket if not)

**Comment by ilijabojanovic** [comment]
Verified

**Comment by bitsofinfo** [comment]
@buger, using dashboard 1.6, now if I take a hashed key that I see in the dashboard (1.6) from "activity by key", then go to "Keys" and put the hash into the search field, check the "this is a hashed key" and still get the error **"Could not retrieve key detail" and "key doesn't exist"**

**Comment by bitsofinfo** [comment]
just realizing this is contingent on the gateways also being 2.6, we've yet to upgrade them. Was trying this via the hybrid dashboard
